### ISSUE TITLE
Unload one Player, load another with same video element

### ISSUE THREAD

**Issue by joeyparrish** [shaka-project/shaka-player#1004]
(Reported through internal channels.)  When one Player instance is unloaded, then another is created on the same video element, this causes an error.  Since the first Player owns the video element until `destroy()` is resolved, the second Player triggers an error when it tries to set up MediaSource on the same video element.

Shaka Player was not designed for this setup.  The video element is owned by the Player instance until destroyed.  Multiple Player instances may exist in parallel, provided that they all use different video elements.

It may be worth writing a short doc explaining the load, unload, and destroy methods, as well as the exclusivity Player needs over the video element.

**Comment by chrisfillmore** [comment]
Could this be enforced programmatically? I am essentially adding this behaviour to our shaka wrapper module right now, so that the client application can't request two players for the same video element.

```
shaka.Player.currentPlayers = new WeakMap();

shaka.Player.getInstance = function (mediaElement) {
  var player = shaka.Player.currentPlayers.get(mediaElement);
  if (player) {
    return currentPlayer;
  }

  player = new shaka.Player(media);
  // You would need to add a 'destroyed' event
  player.addEventListener('destroyed', function () {
    shaka.Player.currentPlayers.delete(mediaElement);
  }

  shaka.Player.currentPlayers.set(mediaElement, player);

  return player;
}
```
Something like the above? Just an idea.

**Comment by joeyparrish** [comment]
Perhaps.  I hadn't planned anything of the sort, but we will mull it over.

**Comment by samanmohamadi** [comment]
will this be resolve in v2.5?
I want to completely destroy the player and recreate it again but after recreating, the player looks like is dead, manifest fetched but player not working.

**Comment by TheModMaker** [comment]
This issue is about documentation, there is nothing to fix here.  After #1087, we should support "detaching" the video from the Player so the video can be used by another instance.

@samanmohamadi If you are trying to "completely" destroy the player, you should be able to do this with `player.destroy`, for example:

```js
oldPlayer.destroy().then(function() {
  newPlayer = new shaka.Player(video);
  // Use newPlayer...
});
```

If you are having trouble with `destroy`, you should file a new bug.

**Comment by joeyparrish** [comment]
You can completely destroy as @TheModMaker points out, or you can detach from the video element and attach to a new one:

```js
player.detach().then(function() {
  // The old video element is no longer referenced by the player,
  // and MediaSource has been torn down.
  return player.attach(newVideo);
});
```

Or in modern browsers:

```js
await player.detach();
// The old video element is no longer referenced by the player,
// and MediaSource has been torn down.
await player.attach(newVideo);
// The player is fully ready for playback,
// and MediaSource has been set up on the new element.
```

**Comment by nargesghahremani** [comment]
hi
i need load video source on same video element, first video load successfully but second one not.
this is my snippet:
```var qplayer;
function initPlayer(manifestUri) {
    // destroy exist player
    qplayer = new shaka.Player();
    var video = document.getElementById('video');
    if(currentIndex == 0) {
      qplayer.attach(video);
    }
      if(qplayer && currentIndex > 0) {
        console.log("destroy old video and create new one in same element")
        qplayer.detach().then(function() {
          // qplayer.unload();
          // qplayer.destroy();
          qplayer.attach(video);
          // The old video element is no longer referenced by the player,
          // and MediaSource has been torn down.
        });
      }
    // Create a Player instance.
    window.qplayer = qplayer;
    // Listen for error events.
    qplayer.addEventListener('error', onErrorEvent);
    // Try to load a manifest.
    // This is an asynchronous process.
    qplayer.load(manifestUri).then(function () {
        // This runs if the asynchronous load is successful.      
    }).catch(onError); // onError is executed if the asynchronous load fails. 
}
```
(i call initPlayer() after user click to see other video)
and get following errors:
Error code 7002 object ,
Error code 3015 object 
i also try this:
```
qplayer.destroy().then(function() {
        qplayer = new shaka.Player(video);          
 });
```
please help:)


@joeyparrish

**Comment by nargesghahremani** [comment]
> hi i need load video source on same video element, first video load successfully but second one not. this is my snippet:
> 
> ```
> function initPlayer(manifestUri) {
>     // destroy exist player
>     qplayer = new shaka.Player();
>     var video = document.getElementById('video');
>     if(currentIndex == 0) {
>       qplayer.attach(video);
>     }
>       if(qplayer && currentIndex > 0) {
>         console.log("destroy old video and create new one in same element")
>         qplayer.detach().then(function() {
>           // qplayer.unload();
>           // qplayer.destroy();
>           qplayer.attach(video);
>           // The old video element is no longer referenced by the player,
>           // and MediaSource has been torn down.
>         });
>       }
>     // Create a Player instance.
>     window.qplayer = qplayer;
>     // Listen for error events.
>     qplayer.addEventListener('error', onErrorEvent);
>     // Try to load a manifest.
>     // This is an asynchronous process.
>     qplayer.load(manifestUri).then(function () {
>         // This runs if the asynchronous load is successful.      
>     }).catch(onError); // onError is executed if the asynchronous load fails. 
> }
> ```
> 
> (i call initPlayer() after user click to see other video) and get following errors: Error code 7002 object , Error code 3015 object i also try this:
> 
> ```
> qplayer.destroy().then(function() {
>         qplayer = new shaka.Player(video);          
>  });
> ```
> 
> please help:)
> 
> @joeyparrish

thanks,problem solved with this snippet code:))))))
(Shaka's instance had to be created once)

‍‍‍‍‍‍‍
```var qplayer;
function initPlayer(manifestUri) {
    var video = document.getElementById('video');
    if(currentIndex == 0) {
      qplayer = new shaka.Player(video);
    }
    window.qplayer = qplayer;
    // Listen for error events.
    qplayer.addEventListener('error', onErrorEvent);
    // Try to load a manifest.
    // This is an asynchronous process.
    qplayer.load(manifestUri).then(function () {
        // This runs if the asynchronous load is successful.
        console.log('The video has now been loaded!');
    }).catch(onError); // onError is executed if the asynchronous load fails. 
}
```

**Comment by avelad** [comment]
I was going to write an FAQ entry about this.
However, on reflection, is something that was asked about once in 2017 really a "frequently" asked question?
I'm already worried that our FAQ might be getting long enough that people might not want to read it, so I think I'll just close this six-year-old issue.
